xcm/xcm-builder/src/pay.rs

```diff
@@ -0,0 +1,99 @@
+// Copyright Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! `PayOverXcm` struct for paying through XCM and getting the status back
+
+use frame_support::traits::{
+	tokens::{Pay, PaymentStatus},
+	Get,
+};
+use sp_std::{marker::PhantomData, vec};
+use xcm::prelude::*;
+use xcm_executor::traits::{Convert, QueryResponseStatus, XcmQueryHandler};
+
+/// Implementation of the `frame_support_traits::tokens::Pay` trait, to allow
+/// for generic payments of a given `AssetKind` and `Balance` from an implied origin, to a
+/// beneficiary via XCM, relying on the XCM `TransferAsset` instruction.
+///
+/// `PayOverXcm::pay` is asynchronous, and returns a `QueryId` which can then be used in
+/// `check_payment` to check the status of the XCM transaction.
+///
+pub struct PayOverXcm<
+	DestChain,
+	Router,
+	Querier,
+	BlockNumber,
+	Timeout,
+	AccountId,
+	AccountIdConverter,
+>(PhantomData<(DestChain, Router, Querier, BlockNumber, Timeout, AccountId, AccountIdConverter)>);
+impl<
+		DestChain: Get<xcm::v3::MultiLocation>,
+		Router: SendXcm,
+		Querier: XcmQueryHandler,
+		BlockNumber,
+		Timeout: Get<Querier::BlockNumber>,
+		AccountId: Clone,
+		AccountIdConverter: Convert<MultiLocation, AccountId>,
+	> Pay
+	for PayOverXcm<DestChain, Router, Querier, BlockNumber, Timeout, AccountId, AccountIdConverter>
+{
+	type Beneficiary = AccountId;
+	type AssetKind = xcm::v3::AssetId;
+	type Balance = u128;
+	type Id = Querier::QueryId;
+
+	fn pay(
+		who: &Self::Beneficiary,
+		asset_kind: Self::AssetKind,
+		amount: Self::Balance,
+	) -> Result<Self::Id, ()> {
+		let beneficiary = AccountIdConverter::reverse(who.clone()).map_err(|_| ())?;
+		let mut message = Xcm(vec![
+			UnpaidExecution { weight_limit: Unlimited, check_origin: None },
+			TransferAsset {
+				beneficiary,
+				assets: vec![MultiAsset { id: asset_kind, fun: Fungibility::Fungible(amount) }]
```

Comment:
We also need AssetKind converter to AssetId

Reply:
Won't it always be the same?